src/AggregateFunctions/AggregateFunctionDeltaSumTimestamp.h

```diff
@@ -0,0 +1,159 @@
+
+#include <type_traits>
+#include <experimental/type_traits>
+
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+
+#include <Columns/ColumnVector.h>
+#include <DataTypes/DataTypesDecimal.h>
+#include <DataTypes/DataTypesNumber.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+
+
+namespace DB
+{
+
+template <typename ValueType, typename TimestampType>
+struct AggregationFunctionDeltaSumTimestampData
+{
+    ValueType sum = 0;
+    ValueType first = 0;
+    ValueType last = 0;
+    TimestampType first_ts = 0;
+    TimestampType last_ts = 0;
+    bool seen = false;
+};
+
+template <typename ValueType, typename TimestampType>
+class AggregationFunctionDeltaSumTimestamp final
+    : public IAggregateFunctionDataHelper<
+        AggregationFunctionDeltaSumTimestampData<ValueType, TimestampType>,
+        AggregationFunctionDeltaSumTimestamp<ValueType, TimestampType>
+      >
+{
+public:
+    AggregationFunctionDeltaSumTimestamp(const DataTypes & arguments, const Array & params)
+        : IAggregateFunctionDataHelper<
+            AggregationFunctionDeltaSumTimestampData<ValueType, TimestampType>,
+            AggregationFunctionDeltaSumTimestamp<ValueType, TimestampType>
+        >{arguments, params}
+    {}
+
+    AggregationFunctionDeltaSumTimestamp()
+        : IAggregateFunctionDataHelper<
+            AggregationFunctionDeltaSumTimestampData<ValueType, TimestampType>,
+            AggregationFunctionDeltaSumTimestamp<ValueType, TimestampType>
+        >{}
+    {}
+
+    String getName() const override { return "deltaSumTimestamp"; }
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeNumber<ValueType>>(); }
+
+    void NO_SANITIZE_UNDEFINED ALWAYS_INLINE add(AggregateDataPtr __restrict place, const IColumn ** columns, size_t row_num, Arena *) const override
+    {
+        auto value = assert_cast<const ColumnVector<ValueType> &>(*columns[0]).getData()[row_num];
+        auto ts = assert_cast<const ColumnVector<TimestampType> &>(*columns[1]).getData()[row_num];
+
+        if ((this->data(place).last < value) && this->data(place).seen)
+        {
+            this->data(place).sum += (value - this->data(place).last);
+        }
+
+        this->data(place).last = value;
+        this->data(place).last_ts = ts;
+
+        if (!this->data(place).seen)
+        {
+            this->data(place).first = value;
+            this->data(place).seen = true;
+            this->data(place).first_ts = ts;
+        }
+    }
+
+    // before returns true if lhs is before rhs or false if it is not or can't be determined
+    bool ALWAYS_INLINE before (
+        const AggregationFunctionDeltaSumTimestampData<ValueType, TimestampType> * lhs,
+        const AggregationFunctionDeltaSumTimestampData<ValueType, TimestampType> * rhs
+    ) const
+    {
+        if (lhs->last_ts < rhs->first_ts)
+        {
+            return true;
+        }
+        if (lhs->last_ts == rhs->first_ts && lhs->last_ts < rhs->last_ts)
```

Comment:
If we assume that we can have one timestamp multiple times, this code will say that (first_ts=4, last_ts=5) is not before (first_ts=5, last_ts=5). Maybe checking `lhs->last_ts <= rhs->first_ts` will be enough in `before`?

Reply:
This is a good point. However I think that my implementation of `before` is correct, but produces nondeterministic results for equal timestamps. Your suggestion would also produce nondeterministic results. 

I think what we actually need to do is handle the case where the timestamps are equal by comparing the values. We could do this in a final `else` clause of the large `if` statement in `merge`. We would want to either always take the smaller value or the larger value into a new 'singleton' state, where a singleton state is a state where first_ts = last_ts and first = last.

Whether we take the smaller or larger, there is a case where it will lose information, but the sum returned would be deterministic. In particular when there are counter resets around where we have two values with the same timestamp.

If we were to implement before with a simple <=, consider the following singleton states:

```
ts | val
0  | 0
1  | 2
1  | 5
2  | 7
```

If we happen to be merging the (ts=1, val=5) state into the (ts=1, val=2) state, `before` would return true (1 <= 1 is true), so we would say that the state with value 2 happened before the state with value 5. This would produce a new state where (first_ts=1, last_ts=1, first=2, last=5, sum=3). So then we're merging (ts=0, val=0) (first_ts=1, last_ts=1, first=2, last=5, sum=3) (ts=2, val=7). This produces a delta sum of 7.

However, if we happen to merge (ts=1, val=2) into (ts=1, val=5), `before` is still true so we end up with (first_ts=1, last_ts=1, first=5, last=2, sum=0). Then we're merging (ts=0, val=0) (first_ts=1, last_ts=1, first=5, last=2, sum=0) (ts=2, val=7). This produces a delta sum of 10.

So the <= approach is nondeterministic.

I will put up my solution and you can see if it makes sense.